Configurable Authentication Process

codinfox opened the pull request:
增加可选的身份验证，增加了对于Google OAuth2.0的支持，同时过滤非@baixing.com/net的邮箱地址。
## 本PR所作工作
1. 对于目录结构进行了整理，将config.[sample.]json和config.js从data/移动到了config/目录下，同时在server/下创建auth/目录，用来几种存放登陆所使用的adapters/rules/和配置用的模块
2. `authentication-*`命名方式的文件均为针对不同API的adapter，以`authentication-google-oauth2`为例：
   
   ``` js
   // Adapter for Google OAuth 2.0 API
   module.exports = function(method) {
   return {
   // 策略在passport.js当中对应的名称，查阅passport.js文档
       strategyName: 'google', 
   // 与登录API提供平台有关，可参阅passport.js对应插件（如passport-google-auth的文档）
       strategyConfiguration: { ... }, 
   // 主要用于登录时指定API所需要的scope等信息，可参阅passport.js对应插件（如passport-google-auth的文档）
       authenticationConfiguration: { ... }, 
   // 对应的passport策略
       strategy: require('passport-google-oauth').OAuth2Strategy 
   };
   };
   ```
   
   这部分的内容对于用户透明
3. `setup-authentication.js`将会自动读取config.json，所有在config.json当中被指定的策略都会被用于登录验证
4. config.json的格式进行了更新，可以参见`config.sample.json`：
   
   ``` json
   {
   "dbfile": "data/db",
   "port": 3000,
   "baseurl": "<Root path of the project>",
   "auth": {
     "emailpattern": "<Your Email Pattern>",
     "methods": {
       "google-oauth2": {
         "enabled": <true or false>,
         "clientID": "<Your Client ID>",
         "clientSecret": "<Your Secret>"
       }
     }
   }
   }
   ```
   
   需要注意的是：
   1. 如果`enabled`字段指定为非true的任何值，下面的配置都将被忽略
   2. `methods`下面的字段，如`google-oauth2`一定要与对应的adapter命名相同。如：`google-oauth2`对应的adapter就是`authentication-google-oauth2.js`，否则无法加载
   3. 如果所有的methods的enable项都设置为`false`，或者干脆没有`auth`这个项，则登陆验证取消
   4. `emailpattern`为可选项，可以不出现在config.json中。如果在`auth`下没有`emailpattern`，则通过open auth验证之后，直接登入系统；如果设置了`emailpattern`，会验证email是否符合pattern。该项支持两种格式：
      1. 正则表达式：如`"@baixing\.(com|net)$"`
      2. 数组：如`["baixing.com", "baixing.net"]`
## Work in Progress
- [x] 增加对Google OAuth2 支持
- [x] 增加登出
- [ ] 改善登陆排版
- [x] 增加登录失败时的错误信息
- [x] 将身份验证作为可配置的选项
- [x] 编写readme

close #1

Commit:
Change directory organization and add dependencies

Commit:
Added passport config

Commit:
Support for Google OAuth2

Commit:
Add @baixing.com/net filter

Commit:
Add logout

Commit:
Add error message for Login failure

Commit:
configurable login

Commit:
Hide logout link when unnecessary

Commit:
update readme

Commit:
refactor authentication logic

amio commented:
这个PR说明里最好能概述下你的整体设计，程序结构是怎么组成的，API 约定是什么。

目前这个 TODO 列表比空白要好不少，但是显然你还可以做的更加好 :D

amio commented:
passport 有维持 session 的中间件，express-session 是不是不需要？
https://github.com/jaredhanson/passport#middleware

codinfox commented:
passport的session中间件是为了从session当中恢复用户信息：https://github.com/jaredhanson/passport#sessions 这里我们可以看到一段代码将用户序列化和解序列化，这段代码会被passport.session()中间件调用，用来恢复用户状态。而express-session则是创建和维持session用的。两者必须同时存在才能实现用户登录状态的保持。

Commit:
Refactor to make config.json more sensible

1. Dynamically generate login page according to the authentication
strategies listed in config.json
2. Hide configurations that are unimportant to users

amio commented:
养成好习惯， wip 的 PR 完工后把前缀删掉，然后 @下需要 review 的人。

这样将来其他人来看这个PR，也可以知道这段时间里发生了什么。

amio commented:
你自行测试一下然后就上线吧
